tool/add_missing_c.rb

```diff
@@ -0,0 +1,96 @@
+            "rb_enc_casefold",            # Not found in MRI c
+            "rb_interned_id_p",           # Not found in MRI c
+            "rb_sym_interned_p",          # Not found in MRI c
+            "rb_random_int",              # Not found in MRI c
+            "rb_trap_exec",               # Not found in MRI c
+            "rb_str_buf_new2",            # Defined by macro definition
+            "rb_str_buf_cat2",            # Defined by macro definition
+            "rb_str_dup_frozen",          # Defined by macro definition
+            "rb_hash_uint32",             # Defined by macro definition
+            "rb_hash_uint",               # Defined by macro definition
+            "rb_hash_end",                # Defined by macro definition
+            "rb_check_safe_str",          # Defined by macro definition
+            "rb_w32_times",               # Skip w32 functions
+            "rb_w32_read",                # Skip w32 functions
+            "rb_w32_write",               # Skip w32 functions
+            "rb_w32_write_console",       # Skip w32 functions
+            "rb_w32_asynchronize",        # Skip w32 functions
+            "rb_w32_pow",                 # Skip w32 functions
+            "rb_scan_args",               # TruffleRuby defines this as a macro
+            "rb_funcall",                 # TruffleRuby defines this as a macro
+            "rb_enumeratorize_with_size", # Compilation error when defined
+          ]
+
+header_files.each do |f|
+    File.open(f).each_line do |line|
+      m = line.match(/^(?!#define )\b(\w+) (\w+)\(/)
+      if m && m[2].start_with?("rb_") && !excludes.include?(m[2])
+        header_definitions << "#{m[1]} #{m[2]}"
+      end
+    end
+end
+
+found_definitions = []
+File.open("src/main/c/cext/ruby.c").each_line do |line|
+  m = line.match(/\b(\w+) (\w+)\((.*?)\)/)
+  if m
+    found_definitions << "#{m[1]} #{m[2]}"
+  end
+end
+
+ruby_c_definitions = {}
+ruby_c_files.each do |f|
+    data = IO.binread(f)
+    data.scan(/\b(\w+)\n(\w+)\(([^;]+?)\)\n/m) do |x,y,z|
+        ruby_c_definitions["#{x} #{y}"] = {:full => "#{x} #{y}(#{z})", :name => y}        
+    end
+    # encoding.c defines a few functions differently
+    data.scan(/\b(\w+) (\w+)\((.*?)\) {/) do |x,y,z|
+      ruby_c_definitions["#{x} #{y}"] = {:full => "#{x} #{y}(#{z})", :name => y}        
+    end
+end
+
+missing = header_definitions.select { |d| !found_definitions.include?(d) }
+missing.uniq!
+
+# Some C sources are defined using macros then undefined just before C definition
+prepend_undef = ["rb_enc_str_new_cstr","rb_enc_code_to_mbclen","rb_utf8_str_new",
+                 "rb_utf8_str_new_cstr","rb_str_cat_cstr"]
+
+missing.each do |m|
+  puts ""
+  if prepend_undef.include?(ruby_c_definitions[m][:name])
+    puts "#undef #{ruby_c_definitions[m][:name]}"
+  end
+  definition = ruby_c_definitions[m][:full]
+  definition.gsub!(/[\n\t]/,' ')
+  definition = definition.split(' ').join(' ')
```

Comment:
Minor: `.squeeze(" ")` might be nice here instead of split+join.

Reply:
Fixed.